Product : Retrieving related object with single request

RomainJufer opened the pull request:
This feature is related to the discussion at : https://github.com/DefectDojo/django-DefectDojo/issues/3018

Add a new query parameter to the endpoints for product retrieval (list/retrieve) that allow the automatic expansion of the relational fields. More specifically, the endpoints can now be called with :

- `/products/?expand={array_of_fields_to_expand}`
- `/products/{id}/?expand={array_of_fields_to_expand}`

The fields to expand are provided as a comma-separated array of strings. Upon receiving the parameters, the `ProductSerializers` will try to follow the relations of the fields specified. If a field is not a relation or not existent in the model, it will simply be ignored without error to preserve the original behavior of the endpoint.

As an example, a `Product` is related to a set of regulations and a user for the technical contact. Both of these fields can be expanded by calling the endpoint

`/products/?expand=regulations,technical_contact`

This new parameter has been documented to be included in the swagger documentation.

valentijnscholten commented:
Looks great! Could you add / modify some test cases to make sure the test suite touches these expansion filters? Ideally a test with and without and verify the response on presence/absence.

xens commented on dojo/api_v2/views.py:
```diff
@@ -557,7 +559,20 @@ class DojoMetaViewSet(mixins.ListModelMixin,
+        openapi.Parameter(
+            name="expand", 
+            in_=openapi.IN_QUERY, 
+            description="Fields to expand based on their relation", 
```
Really cool to be able to document it this way, nice finding ! We could give more info on the `description` as it might be hard to understand without context but I'm unable to find something better ^^.

RomainJufer commented:
Yes of course, however, I've got some troubles finding where to put those tests. I was thinking adding them in `tests/Product_unit_test.py` however it seems that these tests go through the UI. In our case, I think it would make more sense to directly test the endpoint since the UI does not use the newly added field expansion mechanism. Therefore, do I put them inside that same file anyway or do I put them somewhere else ? Unfortunately, I didn't manage to find a test suite that tests the API endpoints, do you have one ?

valentijnscholten commented:
`test_rest_framework.py`

Maffooch commented:
Please rebase to the dev branch.

valentijnscholten commented:
I changed the base branch for the PR which seems to work more or less, but there are conflucts so a merge of dev or rebase onto dev is needed.

RomainJufer commented:
I've rebased and added unit tests for the feature. I also improved a little bit the swagger doc by restricting the range of the value taken by the new parameter with an `enum`. My only concern is that I've done that dynamically and by doing so it breaks the encapsulation principle as `_expansion_serializers`should not be accessed from outside. However, since it is for documentation purposes I think it might be fine. What do you think ?

Also, it looks like the flake8 for unit tests is not going through and one of the reasons is
`'dojo.api_v2.views.SonarqubeProductViewSet' imported but unused (F401)`

But since it is not something that I've introduced (I think) I don't know if I should remove it or not ?

valentijnscholten commented:
yeah remove that unused import. and could you resolve the conflicts (and other flake8 items).
